x/storage/types/message.go

```diff
@@ -1532,3 +1535,56 @@ func (msg *MsgRenewGroupMember) ValidateBasic() error {
 
 	return nil
 }
+
+func NewMsgSetTag(operator sdk.AccAddress, resource string, tags *ResourceTags) *MsgSetTag {
+	return &MsgSetTag{
+		Operator: operator.String(),
+		Resource: resource,
+		Tags:     tags,
+	}
+}
+
+// Route implements the sdk.Msg interface.
+func (msg *MsgSetTag) Route() string {
+	return RouterKey
+}
+
+// Type implements the sdk.Msg interface.
+func (msg *MsgSetTag) Type() string {
+	return TypeMsgSetTag
+}
+
+// GetSigners implements the sdk.Msg interface.
+func (msg *MsgSetTag) GetSigners() []sdk.AccAddress {
+	creator, err := sdk.AccAddressFromHexUnsafe(msg.Operator)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{creator}
+}
+
+// GetSignBytes returns the message bytes to sign over.
+func (msg *MsgSetTag) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// ValidateBasic implements the sdk.Msg interface.
+func (msg *MsgSetTag) ValidateBasic() error {
+	_, err := sdk.AccAddressFromHexUnsafe(msg.Operator)
+	if err != nil {
+		return errors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
+	}
+
+	var grn grn2.GRN
+	err = grn.ParseFromString(msg.Resource, true)
+	if err != nil {
+		return errors.Wrapf(gnfderrors.ErrInvalidGRN, "invalid greenfield resource name (%s)", err)
+	}
+
+	if len(msg.Tags.GetTags()) > MaxTagCount {
```

Comment:
we need limit the length of key and value

Reply:
I check that in `ValidateBasic`

Reply:
I mean length of k/v, not length of `tags`